File: dynamic_listings/lua/default/design_patterns/singleton_lazyload.lua
local function LazySingleton(superclass)
    local o = {}
    o.__index = o
    setmetatable(o, superclass)

    function o.new()
        -- We overwrite the constructor to something that returns
        -- the instance if already existing
        if o._instance then
            return o._instance
        end

        local instance = setmetatable({}, o)

        -- Multi-threading: manage race conditions
        -- ----- Critical region start -----
        o._instance = instance
        return o._instance
        -- ----- Critical region end -----
    end

    return o
end

File: dynamic_listings/C++/default/design_patterns/singleton_lazyload.cpp
class LazySingleton {

    private:
        static LazySingleton* instance = nullptr;
        LazySingleton() {}

    public:
        static Singleton getInstance() {
            // Multi-threading: manage race conditions
            // ----- Critical region start -----
            if (instance == nullptr) {
                instance = new Singleton();
            }
            // ----- Critical region end -----

            return *instance;
        }
};
